[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106806
* #106805
* __->__ #106642

Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils

Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -0,0 +1,506 @@
+from torch.ao.quantization.pt2e.graph_utils import find_sequential_partitions
+from torch.ao.quantization.quantizer import QuantizationSpecBase, SharedQuantizationSpec
+
+# TODO: move these to this file
```
should we still have this TODO? Those seem like pretty general quantizer util functions

jerryzh168 replied:
we do we want to move some of them to this file, see other PRs in the stack

andrewor14 commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -0,0 +1,506 @@
+from torch.fx import Node
+from torch.fx.passes.utils.source_matcher_utils import get_source_partitions
+
+# TODO: move QuantizationConfig here
```
Same here. I feel we shouldn't move QuantizationConfig here

jerryzh168 committed:
Update on "[quant][pt2e] Move annotate functions in XNNPACKQuantizer to utils"


Summary:
This is to allow sharing these annotate functions by other quantizers so that writing a new quantizer is easier

note that these annotation functions will be maintained by XNNPACKQuantizer developers instead of AO team

Test Plan:
python test/test_quantization.py TestQuantizePT2E

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/quantizer/xnnpack_quantizer_utils.py:
```diff
@@ -0,0 +1,510 @@
+        )
+
+
+OP_TO_ANNOTATOR = {
```
In a follow up diff can you make use of decorator pattern to register annotators?

Another thing is that some of the annotate functions are associated with sequential patterns. e.g. Conv2d_relu. Annotate function for that uses find_sequential_partitions. I am wondering if it makes sense to use that as key of the dict.

e.g. instead of OP_TO_ANNOTATOR["conv2d_relu"] = annotate_fn, it could be
OP_TO_ANNOTATOR[[nn.Conv2d, nn.Relu]] = annotate_conv2d_relu_fn
OP_TO_ANNOTATOR[[nn.Conv2d, functional.Relu]] = annotate_conv2d_relu_fn

and registration decorator can enumerate all possible ways of expressing conv2d + relu pattern (e.g. module x functional)

jerryzh168 replied:
sure I can use decorators.

for the other suggestions:
```
OP_TO_ANNOTATOR[[nn.Conv2d, nn.Relu]] = annotate_conv2d_relu_fn
```
right now when we use it, we'll call OP_TO_ANNOTATOR["conv2d_relu"] and it will annotate all conv2d relu patterns, if we change the key, are you suggesting to change the use to
```
OP_TO_ANNOTATOR[(nn.Conv2d, nn.ReLU)](...)
OP_TO_ANNOTATOR[(nn.Conv2d, functional.relu)](...)
```
while all these calls are doing the same thing?

the implementation of `annotate_conv2d_relu` is already tied to the patterns, so I feel it probably does not make a lot of sense to use this directly, maybe I can add another map to store this information if you want?

like 
```
"conv2d_relu": [[nn.Conv2d, nn.ReLU], [nn.Conv2d, functional.relu]]
```

jerryzh168 commented:
@pytorchbot merge
